### PR TITLE
chore: move `net2` dependency out of tendermint crate

### LINKED ISSUE #289: Feature guard the networking stuff
- since we already implemented the networking stuff ourselves, so it would be good for us to cut some dependencies.
- we can even try to make it `no_std`, after separating out networking features.

### PR DESCRIPTION
**Pull request by xla**
This is isolates the code paths which depend on network and http functionality by moving the rpc module into its own crate (tendermint-rpc). The expected outcome is that third-parties which depend on the tendermint crate are not forced to pull in unnecessary dependencies. Furthermore it will isolate any potential crates used to drive an rpc server implementation going forward.

~~Controversially this also removes the dependency on `tokio` as it was mostly used for async test execution, which was achieved by making extensive use of `futures::executor::block_on` - we've come full circle. Eager to hear opinions and alternatives to that.~~
Update: now a dev-dependency only instead of the above.

Follow-up to #7 
Preparation for #219
Closes #289

***

* [x] Referenced an issue explaining the need for the change
* [x] Updated all relevant documentation in docs
* [x] Updated all code comments where relevant
* [ ] ~~Wrote tests~~
* [ ] Updated CHANGES.md

### CODE REVIEW

File: rpc/src/endpoint/block.rs [hunk 1]
```diff
@@ -1,11 +1,9 @@
-//! `/block` endpoint JSONRPC wrapper
+//! `/block` endpoint JSONcrate wrapper
```
Comment (Liamsi) [inline rpc/src/endpoint/block.rs:1]: This looks like an accidental change? It's still a jsonrpc wrapper.

File: rpc/src/endpoint/block_results.rs [hunk 2]
```diff
@@ -1,8 +1,9 @@
-//! `/block_results` endpoint JSONRPC wrapper
+//! `/block_results` endpoint JSONcrate wrapper
```
Comment (Liamsi) [inline rpc/src/endpoint/block_results.rs:1]: same as above

File: rpc/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,20 @@
+[package]
+name = "tendermint-rpc"
```
Comment (Liamsi) [inline rpc/Cargo.toml:2]: 👍

File: rpc/src/endpoint/abci_query.rs [hunk 4]
```diff
@@ -1,13 +1,12 @@
...
-    block,
-    merkle::proof::Proof,
-    rpc, serializers,
-};
 use serde::{Deserialize, Serialize};
 
+use tendermint::abci::{Code, Log, Path};
+use tendermint::block;
+use tendermint::merkle::proof::Proof;
+use tendermint::serializers;
```
Comment (Liamsi) [inline rpc/src/endpoint/abci_query.rs:8]: Just out of curiosity: are all these import changes due to a different formatter? Is there a re reason why all the `use crate:: ...`s are replaced in that fashion?

File: rpc/src/endpoint/blockchain.rs [hunk 5]
```diff
@@ -1,9 +1,10 @@
-//! `/block` endpoint JSONRPC wrapper
+//! `/block` endpoint JSONcrate wrapper
```
Comment (Liamsi) [inline rpc/src/endpoint/blockchain.rs:1]: same

File: rpc/src/endpoint/broadcast/tx_async.rs [hunk 6]
```diff
@@ -14,17 +12,17 @@ pub struct Request {
 }
 
 impl Request {
-    /// Create a new async transaction broadcast RPC request
+    /// Create a new async transaction broadcast crate request
```
Comment (Liamsi) [inline rpc/src/endpoint/broadcast/tx_async.rs:15]:
```suggestion
    /// Create a new async transaction broadcast RPC request
```

File: rpc/src/endpoint/broadcast/tx_sync.rs [hunk 7]
```diff
@@ -20,11 +18,11 @@ impl Request {
...
 }
 
-impl rpc::Request for Request {
+impl crate::Request for Request {
     type Response = Response;
 
-    fn method(&self) -> rpc::Method {
-        rpc::Method::BroadcastTxSync
+    fn method(&self) -> crate::Method {
+        crate::Method::BroadcastTxSync
```
Comment (Liamsi) [inline rpc/src/endpoint/broadcast/tx_sync.rs:25]: Couldn't all these be simply `Method::BroadcastTxSync` etc now that they are in a separate crate?

File: rpc/src/lib.rs [hunk 8]
```diff
@@ -1,11 +1,9 @@
-//! Tendermint RPC: JSONRPC over HTTP support
-//!
-//! Wraps the RPC API described at: <https://tendermint.com/rpc/>
+//! Tendermint RPC definitons and types.
 
-mod client;
-pub mod endpoint;
 pub mod error;
 // TODO(ismail): document fields or re-use the abci types
```
Comment (Liamsi) [inline rpc/src/lib.rs:4]: This comment belongs to the `#[allow(missing_docs)]` below.

File: tendermint/Cargo.toml [hunk 9]
```diff
@@ -55,11 +52,8 @@ subtle-encoding = { version = "0.5", features = ["bech32-preview"] }
...
 toml = { version = "0.5" }
-uuid = { version = "0.8", default-features = false }
 zeroize = { version = "1.1", features = ["zeroize_derive"] }
-async-tungstenite = {version="0.5", features = ["tokio-runtime"]}
-tokio = { version = "0.2", features = ["macros"] }
 ripemd160 = "0.9"
 
 [dev-dependencies]
-serde_json = "1"
+tendermint-rpc = { version = "0.1.0", path = "../rpc" }
```
Comment (Liamsi) [inline tendermint/Cargo.toml:59]: Nice, nice.

File: tendermint/src/lite/verifier.rs [hunk 10]
```diff
@@ -528,22 +528,21 @@ mod tests {
...
+    fn assert_bisection_ok(
         req: &MockRequester,
         ts: &TrustedState<MockCommit, MockHeader>,
         untrusted_height: u64,
         expected_num_of_requests: usize,
         expected_final_state: &MockState,
     ) {
         let mut cache: Vec<MockTrustedState> = Vec::new();
-        let ts_new = verify_bisection_inner(
+        let ts_new = futures::executor::block_on(verify_bisection_inner(
```
Comment (Liamsi) [inline tendermint/src/lite/verifier.rs:539]: is this necessary to get rid of the direct tokio dependency?

File: tendermint/src/lite/verifier.rs [hunk 11]
```diff
@@ -784,20 +782,18 @@ mod tests {
         assert_single_err(ts, vac, err.into());
     }
 
-    #[tokio::test]
-    async fn test_verify_bisection_1_val() {
```
Comment (Liamsi) [inline tendermint/src/lite/verifier.rs:788]: Ok, answers my question I guess :-D 
https://github.com/informalsystems/tendermint-rs/pull/338/files#r441736094

File: tendermint/src/serializers.rs [hunk 12]
```diff
@@ -50,5 +50,5 @@ pub(crate) use raw_commit_sig::RawCommitSig;
 mod tests;
 
 mod custom;
-pub(crate) use custom::parse_non_empty_block_id;
-pub(crate) use custom::parse_non_empty_hash;
+pub use custom::parse_non_empty_block_id;
+pub use custom::parse_non_empty_hash;
```
Comment (Liamsi) [inline tendermint/src/serializers.rs:54]: Do we have to publish these beyond the crate? Probably, yes as some might be used in the rpc crate now?

File: rpc/Cargo.toml (hunk 3 above)
Comment (Liamsi) [inline rpc/Cargo.toml:2]: This probably deserves a short readme (can be done in a followup PR).

File: rpc/Cargo.toml [hunk 13]
```diff
@@ -0,0 +1,20 @@
...
+async-tungstenite = {version="0.5", features = ["tokio-runtime"]}
+bytes = "0.5"
+futures = "0.3"
+getrandom = "0.1"
+http = "0.2"
+hyper = "0.13"
+serde = { version = "1", features = [ "derive" ] }
+serde_bytes = "0.11"
+serde_json = "1"
+tendermint = { version = "0.13.0", path = "../tendermint" }
```
Comment (Liamsi) [inline rpc/Cargo.toml:17]: I'm always sceptical of circular dependencies (even if it's just a dev-dependency). Are we confident that this does cannot have weird side-effects?

File: rpc/src/endpoint/abci_query.rs (hunk 4 above)
Comment (xla) [reply rpc/src/endpoint/abci_query.rs:8]: Personal touch really, optimised for editability and tracebility of dependencies. I usually one liner per top-level module in a crate.

File: tendermint/src/lite/verifier.rs (hunk 11 above)
Comment (Liamsi) [reply tendermint/src/lite/verifier.rs:788]: Also, I should better read the introductory comments 🙈 😆 

> Controversially this also removes the dependency on tokio as it was mostly used for async test execution, which was achieved by making extensive use of futures::executor::block_on

File: rpc/Cargo.toml (hunk 13 above)
Comment (xla) [reply rpc/Cargo.toml:17]: I would not expect any, but truth be told this was the solution with the least amount of opinionated changes. Would like to get rid of this circle as well, but afraid this goes too far for this PR.

File: rpc/src/endpoint/broadcast/tx_sync.rs (hunk 7 above)
Comment (xla) [reply rpc/src/endpoint/broadcast/tx_sync.rs:25]: We could, but there are some reoccurring names here `Request`, `Response`. Wanna keep it clear where all of them come from.

**Comment by xla** [comment]
> But more relevant is that some of the changes seem to have messed with the integration tests:
> `thread 'rpc::abci_info' panicked at 'there is no reactor running, must be called from the context of Tokio runtime',`

On second thought it might be worthwhile to keep tokio in `dev-dependencies`.

File: tendermint/src/serializers.rs (hunk 12 above)
Comment (xla) [reply tendermint/src/serializers.rs:54]: Indeed, the rpc crate depends on a lot of these serialisers. Not happy with it. Also found the excessive use of `pub(crate)` is unnecessary as the top-level lib determines visibility beyond crate boundaries. Eager to hear how we can address this issue.

Comment (Liamsi) [reply tendermint/src/serializers.rs:54]: If they are *only* used in the rpc lib, they could actually live there. Not sure yet.

File: rpc/src/endpoint/abci_query.rs (hunk 4 above)
Comment (Liamsi) [reply rpc/src/endpoint/abci_query.rs:8]: Ok for me. Ideally, these should be consistent in all our crates though.

File: tendermint/src/serializers.rs (hunk 12 above)
Comment (xla) [reply tendermint/src/serializers.rs:54]: They are also used in `tendermint::block`.

Comment (Liamsi) [reply tendermint/src/serializers.rs:54]: Maybe removing the net dependency outweighs this now publisized types. 

I might have mentioned that somewhere else before but ideally, the serialzation code would simply be abstracted away in the sense that anyone using the tendermint crate can e.g. simply call block::deserialize (or deserialize_json and deserialize_binary) and the that could internally use a dedicated serialization type serialize::block (and impl Froms between both) which uses methods / helpers like the above. Anyone (like the rpc crate) could simply use the core types and wouldn't need to care about serialization specific stuff. Does that make sense? Probably worth another discussion and later another PR.

File: rpc/Cargo.toml (hunk 13 above)
Comment (Liamsi) [reply rpc/Cargo.toml:17]: I can't think of any problems either 🍀 🤞

File: tendermint/src/serializers.rs (hunk 12 above)
Comment (Liamsi) [reply tendermint/src/serializers.rs:54]: For now: maybe the `serializers` module should include a comment that we don't guarantee any backwards compatibility on the methods of that module and consider them an internal detail (use at you own risk alike).

File: tendermint/src/serializers.rs [hunk 14]
```diff
@@ -2,6 +2,10 @@
 //!
 //! Serializers and deserializers for a transparent developer experience.
 //!
+//! CAUTION: There are no guarantees for backwards compatibility, this module should be considered
+//! an internal implementation detail which can vanish without further warning. Use at your own
+//! risk.
```
Comment (Liamsi) [inline tendermint/src/serializers.rs:7]: 👍

File: tendermint/src/serializers.rs (hunk 12 above)
Comment (ebuchman) [reply tendermint/src/serializers.rs:58]: We need to lay out an explicit semantic versioning scheme and declare which modules we're covering and which we're not. Good to start tracking like this though thanks!

**Pull request merged**